File: hnswlib-core/src/main/java/org/github/jelmerk/knn/SearchResult.java
package org.github.jelmerk.knn;

import java.io.Serializable;
import java.util.Objects;

/**
 * Result of a nearest neighbour search.
 *
 * @param <TItem> type of the item returned
 * @param <TDistance> type of the distance returned by the configured distance function
 */
public class SearchResult<TItem, TDistance extends Comparable<TDistance>>
        implements Comparable<SearchResult<TItem, TDistance>>, Serializable {

    private static final long serialVersionUID = 1L;

    private final TDistance distance;

    private final TItem item;

    /**
     * Constructs a new SearchResult instance.
     *
     * @param item the item
     * @param distance the distance from the search query
     */
    public SearchResult( TItem item, TDistance distance) {
        this.item = item;
        this.distance = distance;
    }

    /**
     * Returns the item.
     *
     * @return the item
     */
    public TItem item() {
        return item;
    }

    /**
     * Returns the distance from the search query.
     *
     * @return the distance from the search query
     */
    public TDistance distance() {
        return distance;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public int compareTo(SearchResult<TItem, TDistance> o) {
        return this.distance.compareTo(o.distance);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        SearchResult<?, ?> that = (SearchResult<?, ?>) o;
        return Objects.equals(distance, that.distance) &&
                Objects.equals(item, that.item);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public int hashCode() {
        return Objects.hash(distance, item);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public String toString() {
        return "SearchResult{" +
                "distance=" + distance +
                ", item=" + item +
                '}';
    }
}

File: hnswlib-scala/src/main/scala/com/github/jelmerk/hnswlib/scala/SearchResult.scala
package com.github.jelmerk.hnswlib.scala

import com.github.jelmerk.hnswlib.core

import java.util.Comparator

object SearchResult {

  /**
    * SearchResult factory.
    *
    * @param item the item
    * @param distance the distance from the search query
    * @param ordering the ordering
    * @tparam TItem type of the item returned
    * @tparam TDistance type of the distance returned by the configured distance function
    * @return a new SearchResult
    */
  def apply[TItem, TDistance](item: TItem, distance: TDistance)(implicit ordering: Ordering[TDistance])
    : core.SearchResult[TItem, TDistance] =
      new core.SearchResult[TItem, TDistance](item, distance, ordering)

  /**
    * SearchResult factory.
    *
    * @param item the item
    * @param distance the distance from the search query
    * @tparam TItem type of the item returned
    * @tparam TDistance type of the distance returned by the configured distance function
    * @return a new SearchResult
    */
  def apply[TItem, TDistance](item: TItem, distance: TDistance, distanceComparator: Comparator[TDistance])
    : core.SearchResult[TItem, TDistance] =
      new core.SearchResult[TItem, TDistance](item, distance, distanceComparator)

  /**
    * Extract fields from SearchResult.
    *
    * @param result searchresult to extract the fields from
    * @tparam TItem type of the item returned
    * @tparam TDistance type of the distance returned by the configured distance function
    * @return the extracted fields
    */
  def unapply[TItem, TDistance](result: core.SearchResult[TItem, TDistance]): Option[(TItem, TDistance)] =
    Some(result.item -> result.distance)

}
